mundane-bytes/src/main/scala/com/ambiata/mundane/bytes/Buffer.scala

```diff
@@ -0,0 +1,77 @@
+package com.ambiata.mundane.bytes
+
+/**
+ * <pre>
+ *                  uuuuuuu
+ *               uu$$$$$$$$$$$uu
+ *           uu$$$$$$$$$$$$$$$$$uu
+ *          u$$$$$$$$$$$$$$$$$$$$$u
+ *         u$$$$$$$$$$$$$$$$$$$$$$$u
+ *        u$$$$$$$$$$$$$$$$$$$$$$$$$u
+ *        u$$$$$$$$$$$$$$$$$$$$$$$$$u
+ *        u$$$$$$"   "$$$"   "$$$$$$u
+ *        "$$$$"      u$u       $$$$"
+ *         $$$u       u$u       u$$$
+ *         $$$u      u$$$u      u$$$
+ *          "$$$$uu$$$   $$$uu$$$$"
+ *           "$$$$$$$"   "$$$$$$$"
+ *             u$$$$$$$u$$$$$$$u
+ *              u$"$"$"$"$"$"$u
+ *   uuu        $$u$ $ $ $ $u$$       uuu
+ *  u$$$$        $$$$$u$u$u$$$       u$$$$
+ *   $$$$$uu      "$$$$$$$$$"     uu$$$$$$
+ * u$$$$$$$$$$$uu    """""    uuuu$$$$$$$$$$
+ * $$$$"""$$$$$$$$$$uuu   uu$$$$$$$$$"""$$$"
+ *  """      ""$$$$$$$$$$$uu ""$"""
+ *            uuuu ""$$$$$$$$$$uuu
+ *   u$$$uuu$$$$$$$$$uu ""$$$$$$$$$$$uuu$$$
+ *   $$$$$$$$$$""""           ""$$$$$$$$$$$"
+ *    "$$$$$"                      ""$$$$""
+ *      $$$"                         $$$$"
+ * </pre>
+ *
+ * This data type is _not_ intended as a safe, immutable array of bytes like
+ * [[https://github.com/scodec/scodec-bits/blob/master/core/src/main/scala/scodec/bits/ByteVector.scala ByteVector]].
+ * Instead, this is a c-like struct for wrapping an array with an offset/length.
+ * In other words - it's bringing convenience to crazy-town, not correctness.
+ *
+ * The important constraint is that both [[bytes]] and [[offset]] are immutable, and a new copy of the [[Buffer]] needs
+ * to created to resize or offset the array. Consumers can then do an instance check on [[Buffer]] to confirm
+ * whether the wrapped bytes (may) have been modified or whether an extra/final copy needs to be done.
+ */
+case class Buffer private(bytes: Array[Byte], offset: Int, private var _length: Int) {
+
+  def length: Int =
+    _length
+}
+
+object Buffer {
+
+  def empty(initialSize: Int): Buffer =
+    wrapArray(new Array[Byte](initialSize), 0, 0)
+
+  def wrapArray(bytes: Array[Byte], offset: Int, length: Int): Buffer = {
+    if (bytes.length == 0)
+      sys.error("Cannot wrap an empty array")
+    if (offset < 0 || offset >= bytes.length)
+      sys.error(s"Invalid offset $offset for array of length ${bytes.length}")
+    if (length < offset || length > bytes.length)
```

Comment:
> Shouldn't this be: 

Yeah. Sorry I've been thinking of `length` of the `endOffset` which isn't right. Will fix.